system-addon/content-src/components/TopSites/TopSites.jsx

```diff
@@ -69,6 +70,93 @@ class TopSite extends React.Component {
   }
 }
 
+/**
+ * A proxy class that uses double requestAnimationFrame from
+ * componentDidMount to dispatch a SAVE_SESSION_PERF_DATA to the main procsess
+ * after the paint.
+ *
+ * This uses two callbacks because, after one callback, this part of the tree
+ * may have rendered but not yet reflowed.  This strategy is modeled after
+ * https://stackoverflow.com/a/34999925 but uses a double rFA because
+ * we want to get to the closest reliable paint for measuring, and
+ * setTimeout is often throttled or queued by browsers in ways that could
+ * make it lag too long.
+ *
+ * XXX Should probably be split out into a higher-order component to wrap
+ * whatever.
+ *
+ * @class TopSitesPerfTimer
+ * @extends {React.Component}
+ */
+class TopSitesPerfTimer extends React.Component {
+  constructor(props) {
+    super(props);
+    this.perfSvc = this.props.perfSvc || perfSvc;
+
+    this._maybeSendPaintedEvent = this._maybeSendPaintedEvent.bind(this);
+    this._sendPaintedEvent = this._sendPaintedEvent.bind(this);
+    this._timestampSent = false;
+  }
+
+  componentDidMount() {
+    this._maybeSendPaintedEvent();
+  }
+
+  componentDidUpdate() {
+    this._maybeSendPaintedEvent();
+  }
+
+  /**
+   * Call the given callback when the next animation frame (not the upcoming
+   * one) paints.
```

Comment:
> I suppose the confusion comes from the requestAnimationFrame is at the beginning of the frame before it actually happens.

This is indeed exactly the problem.  I've changed the verbiage to include the word subsequent, which maybe helps a bit.